Start porting the CWS templates from Tornado to Jinja2

lerks opened the pull request:
This is the first of a series of pull requests that ports CWS's templates from the Tornado engine to Jinja2. This pull request does many atomic changes that gap the syntax differences between the two formats.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cms-dev/cms/845)
<!-- Reviewable:end -->

stefano-maggiolo commented on cms/server/contest/templates/documentation.html:
```diff
@@ -8,13 +8,13 @@ <h1>{{ _("Documentation") }}</h1>
 <h2>{{ _("Programming languages and libraries") }}</h2>
 
-{% if any(contest.languages|map("get_language")|map(attribute="source_extensions")|select("contains", ".cpp")) %}
+{% if contest.languages|map("get_language")|map(attribute="source_extensions")|select("contains", ".cpp")|list|length > 0 %}
```
Can we define our own? like `|nonempty`?

stefano-maggiolo replied:
I see you defined any later, still nonempty wouldn't be bad for all other occasions you do len() > 0

lerks replied:
As you mentioned, I added `all` and `any` that make some of these clearer. I think the benefit of `|empty` vs `|length == 0` and `|nonempty` vs `|length > 0` (which, I guess, are brevity and supposed better readability) are not worth the introduction of new filters that may confuse anyone not familiar with our custom idiom.

stefano-maggiolo commented on cms/server/jinja2_toolbox.py:
```diff
@@ -32,32 +32,64 @@
+
+
+@contextfilter
+def any_(ctx, l, test=None, *args):
```
Docstring

lerks replied:
Done.

stefano-maggiolo commented on cms/server/jinja2_toolbox.py:
```diff
@@ -32,32 +32,64 @@
-def dictselect(env, d, *args, **kwargs):
-    if len(args) == 0:
+@contextfilter
+def all_(ctx, l, test=None, *args):
```
Docstring

lerks replied:
Done.

lerks commented on cms/server/contest/jinja2_toolbox.py:
```diff
@@ -0,0 +1,42 @@
+CWS_ENVIRONMENT = GLOBAL_ENVIRONMENT.overlay(
+    extensions=['jinja2.ext.i18n'],
+    loader=PackageLoader('cms.server.contest', 'templates'))
+CWS_ENVIRONMENT.policies['ext.i18n.trimmed'] = True
```
Done, added comments to every option.

lerks commented on cms/server/contest/templates/documentation.html:
```diff
@@ -8,13 +8,13 @@ <h1>{{ _("Documentation") }}</h1>
 <h2>{{ _("Programming languages and libraries") }}</h2>
 
-{% if any(".cpp" in get_language(lang).source_extensions for lang in contest.languages) %}
+{% if any(contest.languages|map("get_language")|map(attribute="source_extensions")|select("contains", ".cpp")) %}
```
Agreed that with get_laguage->to_language and the new any filter this becomes clearer.

lerks commented on cms/server/jinja2_toolbox.py:
```diff
@@ -32,7 +32,25 @@
+
+
+@environmentfilter
+def dictselect(env, d, *args, **kwargs):
```
Done.

lerks commented on cms/server/contest/templates/printing.html:
```diff
@@ -44,7 +44,7 @@ <h2 style="margin-bottom: 10px">{{ _("Print") }}</h2>
 <h2 style="margin: 40px 0 10px">{{ _("Previous print jobs") }}</h2>
 
-{% set show_date = any(p.timestamp.replace(tzinfo=utc).astimezone(timezone).date() != now.replace(tzinfo=utc).astimezone(timezone).date() for p in printjobs) %}
+{% set show_date = any(printjobs|map(attribute="timestamp")|map("astimezone", timezone)|map(attribute="date")|map("call")|select("ne", (now|astimezone(timezone)).date())) %}
```
Done.

lerks commented on cms/server/contest/templates/task_description.html:
```diff
@@ -37,7 +37,7 @@ <h2>{{ _("Statement") }}</h2>
         </p>
     {% set task_primary = primary_statements %}
-    {% for statement in sorted(itervalues(task.statements), key=lambda x: x.language_name) %}
+    {% for statement in itervalues(task.statements)|sort(attribute="language_name") %}
```
Not sure what you mean: `task.statement` is a dict where value `s` has key `s.language`, while dictsort can only sort by plain key or value.

lerks commented on cms/server/contest/templates/contest.html:
```diff
@@ -11,7 +11,7 @@
 {% for lang in contest.languages %}
     '{{ lang }}': {
-{% for extension in get_language(lang).source_extensions %}
+{% for extension in (lang|get_language).source_extensions %}
```
Changed to to_language. (PS: it's defined in `cms/server/jinja2_toolbox.py`)

stefano-maggiolo commented:
LGTM, pending moving those few iteritems(x)|sort to dictsort

lerks commented:
Fixed up all commits (with rebasing), tested all again, fixed some bugs, and pushed to the `jinja` branch, which will be merged into master as soon as all commits are in.
